taichi/program/state_flow_graph.cpp

```diff
@@ -15,76 +16,192 @@ TLANG_NAMESPACE_BEGIN
-      [&s](const SFGStateToNodes::value_type &v) { return v.first == s; });
-}
-
-std::pair<SFGStateToNodes::value_type::second_type *, bool> insert(
-    SFGStateToNodes &m,
+std::pair<LatestStateReaders::value_type::second_type *, bool> insert(
+    LatestStateReaders &m,
     const AsyncState &s) {
   int sz = m.size();
   for (int i = 0; i < sz; i++) {
     if (m[i].first == s) {
       return std::make_pair(&m[i].second, true);
     }
   }
-  m.push_back(std::make_pair(s, SFGStateToNodes::value_type::second_type{}));
+  m.push_back(std::make_pair(s, LatestStateReaders::value_type::second_type{}));
   return std::make_pair(&(m.back().second), false);
 }
 
-SFGStateToNodes::value_type::second_type &get_or_insert(SFGStateToNodes &m,
-                                                        const AsyncState &s) {
+LatestStateReaders::value_type::second_type &get_or_insert(
+    LatestStateReaders &m,
+    const AsyncState &s) {
   // get_or_insert() implies that the user doesn't care whether |s| is already
   // in |m|, so we just return the mapped value. This is functionally equivalent
   // to a (unordered) map's operator[].
   return *(insert(m, s).first);
 }
 
-bool insert(SFGStateToNodes &m, const AsyncState &s, StateFlowGraph::Node *n) {
+bool insert(LatestStateReaders &m,
+            const AsyncState &s,
+            StateFlowGraph::Node *n) {
   auto p = insert(m, s);
   const bool b = p.first->insert(n).second;
   return p.second && b;
 }
 
+using StateToNodesMap = StateFlowGraph::StateToNodesMap;
+
 }  // namespace
 // TODO: rename state to edge since we have not only state flow edges but also
 // dependency edges.
 
+StateToNodesMap::StateToNodesMap() {
+  // TODO(mask): need to resize |data_| to |kNumInlined| first?
+}
+void StateToNodesMap::clear() {
+  sorted_ = false;
+  // It seems that once we are beyond the small buffer threshold, there's no
+  // going back...
+  data_.clear();
+}
+
+bool StateToNodesMap::has_edge(const Edge &e) const {
+  // TODO(mask): Could be a bit tricky once mask is implemented: we need to
+  // support both sorted and unsorted cases.
+  // Alternatively, only use this in sorted case, and don't call this inside
+  // insert_edge()?
+  if (!sorted_) {
+    // Shall we do linear search even when sorted if the size is small?
+    // It's gonna be sad if std::find turns out to hurt the performance...
+    return std::find(data_.begin(), data_.end(), e) != data_.end();
+  }
+  return std::binary_search(data_.begin(), data_.end(), e);
+}
+
+void StateToNodesMap::insert_edge(const AsyncState &as, Node *n) {
+  TI_ASSERT(!sorted_);
+  const Edge e = std::make_pair(as, n);
+  if (has_edge(e))
+    return;
+
+  // TODO(mask): check if we should write to the inlined region.
+  data_.push_back(e);
+}
+
+void StateToNodesMap::remove_node(const Node *n) {
+  // This should be used only after |data_| is sorted.
+  // 1) disconnect_with
+  // 2) fuse_range
+  // So adding the mask shouldn't affect this implementation.
```

Comment:
This function doesn't seem to be called very frequently. We can likely use some correct but slow solution.